File: python/ta/typeagent/demo/answer_context_schema.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

from typing import Annotated, Any, Union

from pydantic.dataclasses import dataclass

from ..knowpro.interfaces import DateRange

Doc = lambda s: s  # Placeholder for Doc metadata

EntityNames = Union[str, list[str]]


@dataclass
class RelevantKnowledge:
    knowledge: Annotated[Any, Doc("The actual knowledge")]
    origin: Annotated[
        EntityNames | None, Doc("Entity or entities who mentioned the knowledge")
    ]
    audience: Annotated[
        EntityNames | None,
        Doc("Entity or entities who received or consumed this knowledge"),
    ]
    timeRange: Annotated[
        DateRange | None, Doc("Time period during which this knowledge was gathered")
    ]


@dataclass
class RelevantMessage:
    from_: Annotated[EntityNames | None, Doc("Sender(s) of the message")]
    to: Annotated[EntityNames | None, Doc("Recipient(s) of the message")]
    timestamp: Annotated[str | None, Doc("Timestamp of the message in ISO format")]
    messageText: Annotated[str | list[str] | None, Doc("Text chunks in this message")]


@dataclass
class AnswerContext:
    entities: Annotated[
        list[RelevantKnowledge] | None,
        Doc(
            "Relevant entities. Use the 'name' and 'type' properties of entities to PRECISELY identify those that answer the user question."
        ),
    ]
    topics: Annotated[list[RelevantKnowledge] | None, Doc("Relevant topics")]
    messages: Annotated[list[RelevantMessage] | None, Doc("Relevant messages")]

File: dotnet/typeagent/src/knowpro/Answer/AnswerContextSchema.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace TypeAgent.KnowPro.Answer;

public partial class RelevantKnowledge
{
    // Entity or entities who mentioned the knowledge
    [JsonPropertyName("origin")]
    public OneOrManyItem<string>? Origin { get; set; }

    // Entity or entities who received or consumed this knowledge
    [JsonPropertyName("audience")]
    public OneOrManyItem<string>? Audience { get; set; }

    // Time period during which this knowledge was gathered
    [JsonPropertyName("timeRange")]
    public TimestampRange? TimeRange { get; set; }
};

public partial class RelevantTopic : RelevantKnowledge
{
    [JsonPropertyName("knowledge")]
    public string? Topic { get; set; }
}

public partial class RelevantEntity : RelevantKnowledge
{
    [JsonPropertyName("knowledge")]
    public ConcreteEntity? Entity { get; set; }
}

public partial class RelevantMessage
{
    [JsonPropertyName("from")]
    public OneOrManyItem<string>? From { get; set; }

    [JsonPropertyName("to")]
    public OneOrManyItem<string>? To { get; set; }

    [JsonPropertyName("timestamp")]
    public string? Timestamp { get; set; }

    [JsonPropertyName("messageText")]
    public OneOrManyItem<string>? MessageText { get; set; }
}

public partial class AnswerContext
{
    // Relevant entities
    // Use the 'name' and 'type' properties of entities to PRECISELY identify those that answer the user question.
    public IList<RelevantEntity>? Entities { get; set; }

    // Relevant topics
    public IList<RelevantTopic> Topics { get; set; }

    // Relevant messages
    public IList<RelevantMessage>? Messages { get; set; }
};
